dpcpp/solver/batch_bicgstab_kernels.hpp.inc

```diff
@@ -0,0 +1,449 @@
+         r += item_ct1.get_local_range().size()) {
+        s_shared_entry[r] = r_shared_entry[r] - alpha * v_shared_entry[r];
+    }
+}
+
+
+template <const bool sg_kernel_all, typename ValueType>
+__dpct_inline__ void compute_omega(const int num_rows,
+                                   const ValueType* const t_shared_entry,
+                                   const ValueType* const s_shared_entry,
+                                   ValueType& temp, ValueType& omega,
+                                   sycl::nd_item<3> item_ct1)
+{
+    if constexpr (sg_kernel_all) {
+        auto sg = item_ct1.get_sub_group();
+        const auto sg_id = sg.get_group_id();
+        const auto tid = item_ct1.get_local_linear_id();
+
+        if (sg_id == 0)
+            single_rhs_compute_dot_sg(num_rows, t_shared_entry,
+                                          s_shared_entry, omega, item_ct1);
+        else if (sg_id == 1)
+            single_rhs_compute_dot_sg(num_rows, t_shared_entry,
+                                          t_shared_entry, temp, item_ct1);
+        item_ct1.barrier(sycl::access::fence_space::local_space);
+        if (tid == 0) omega /= temp;
+        item_ct1.barrier(sycl::access::fence_space::local_space);
+    } else {
+        single_rhs_compute_dot(num_rows, t_shared_entry, s_shared_entry,
+                                   omega, item_ct1);
+        single_rhs_compute_dot(num_rows, t_shared_entry, t_shared_entry,
+                                   temp, item_ct1);
+        omega /= temp;
+    }
+}
+
+template <typename ValueType>
+__dpct_inline__ void update_x_and_r(
+    const int num_rows, const ValueType* const p_hat_shared_entry,
+    const ValueType* const s_hat_shared_entry, const ValueType& alpha,
+    const ValueType& omega, const ValueType* const s_shared_entry,
+    const ValueType* const t_shared_entry, ValueType* const x_shared_entry,
+    ValueType* const r_shared_entry, sycl::nd_item<3> item_ct1)
+{
+    for (int r = item_ct1.get_local_linear_id(); r < num_rows;
+         r += item_ct1.get_local_range().size()) {
+        x_shared_entry[r] = x_shared_entry[r] + alpha * p_hat_shared_entry[r] +
+                            omega * s_hat_shared_entry[r];
+        r_shared_entry[r] = s_shared_entry[r] - omega * t_shared_entry[r];
+    }
+}
+
+
+template <typename ValueType>
+__dpct_inline__ void update_x_middle(const int num_rows, const ValueType& alpha,
+                                     const ValueType* const p_hat_shared_entry,
+                                     ValueType* const x_shared_entry,
+                                     sycl::nd_item<3> item_ct1)
+{
+    for (int r = item_ct1.get_local_linear_id(); r < num_rows;
+         r += item_ct1.get_local_range().size()) {
+        x_shared_entry[r] = x_shared_entry[r] + alpha * p_hat_shared_entry[r];
+    }
+}
+
+
+template <typename StopType, const int n_shared_total, const bool sg_kernel_all,
```

Comment:
Is `sg_kernel_all` the same as `prec_shared_bool` in the cuda/hip kernels?
Also, the const is unnecessary here.

Reply:
Also, that name is very non-descriptive.

Reply:
No, it is actually different. With `sg_kernel_all` we create local memory objects that are [shared between all work-items of a group](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_local_memory.asciidoc#allocating-local-memory). I think we observed better performance for smaller group sizes with this strategy.

Reply:
Then, I don't think the current name is appropriate. At least, I could not guess that this is the function of that parameter.

Reply:
maybe `use_sg_mem_scalars`?